edit/util.js

```diff
@@ -0,0 +1,94 @@
+'use strict';
+
+function dirtyReporter() {
+  const dirty = new Map();
+  const onchanges = [];
+
+  function add(obj, value) {
+    const saved = dirty.get(obj);
+    if (!saved) {
+      dirty.set(obj, {type: 'add', newValue: value});
+    } else if (saved.type === 'remove') {
+      if (saved.savedValue === value) {
+        dirty.delete(obj);
+      } else {
+        saved.newValue = value;
+        saved.type = 'modify';
+      }
+    }
+  }
+
+  function remove(obj, value) {
+    const saved = dirty.get(obj);
+    if (!saved) {
+      dirty.set(obj, {type: 'remove', savedValue: value});
+    } else if (saved.type === 'add') {
+      dirty.delete(obj);
+    } else if (saved.type === 'modify') {
+      saved.type = 'remove';
+    }
+  }
+
+  function modify(obj, oldValue, newValue) {
+    const saved = dirty.get(obj);
+    if (!saved) {
+      if (oldValue !== newValue) {
+        dirty.set(obj, {type: 'modify', savedValue: oldValue, newValue});
+      }
+    } else if (saved.type === 'modify') {
+      if (saved.savedValue === newValue) {
+        dirty.delete(obj);
+      } else {
+        saved.newValue = newValue;
+      }
+    } else if (saved.type === 'add') {
+      saved.newValue = newValue;
+    }
+  }
+
+  function clear() {
+    dirty.clear();
+  }
+
+  function isDirty() {
+    return dirty.size > 0;
+  }
+
+  function onChange(cb) {
+    onchanges.push(cb);
+  }
+
+  function wrap(obj) {
+    for (const key of ['add', 'remove', 'modify', 'clear']) {
+      obj[key] = trackChange(obj[key]);
+    }
+    return obj;
+  }
+
+  function emitChange() {
+    for (const cb of onchanges) {
+      try {
```

Comment:
Why try-catch is needed here?

Reply:
I usually use try-catch when invoking event handler so the error won't bubble up to crash the entire app.
```js
dirty.onChange(() => {
  throw new Error();
});
dirty.add('something', 'new value');
// the error from listener won't affect following lines.
doSomethingElse();
```
It's fine to remove it if we don't need such pattern.

Reply:
This pattern may be useful when you write universal code that may be plugged somewhere. But it's meaningless and wrong to blindly add try-catch without investigating whether an error may occur at all, because why not add try-catch around all statements then?

Reply:
It's not a try-catch problem but a sync/async choice. In some environment we can use setImmediate to execute callback, that error won't affect current flow either.

Reply:
What error? There are no errors I know of. If any error pops up, it's better the entire page crashes so that we can fix the actual fault code.